File: python/200.number-of-islands.py
from collections import deque
from itertools import product
from typing import List


class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m, n = len(grid), len(grid[0])

        islands = 0

        for y, x in product(range(m), range(n)):
            if grid[y][x] == "0":
                continue

            islands += 1
            grid[y][x] = "0"
            q = deque([(y, x)])
            while q:
                y, x = q.popleft()
                for dy, dx in ((-1, 0), (1, 0), (0, -1), (0, 1)):
                    ny, nx = y + dy, x + dx
                    if 0 <= ny < m and 0 <= nx < n and grid[ny][nx] == "1":
                        grid[ny][nx] = "0"
                        q.append((ny, nx))
        return islands

File: golang/200.number-of-islands.go
func numIslands(grid [][]byte) int {
    m, n := len(grid), len(grid[0])

    var cleanup func(int, int)
    cleanup = func(i, j int) {
        if i < 0 || j < 0 || i >= m || j >= n || grid[i][j] == '0' {
            return
        }

        grid[i][j] = '0'
        cleanup(i + 1, j)
        cleanup(i, j + 1)
        cleanup(i - 1, j)
        cleanup(i, j - 1)
    }

    islands := 0
    for i := range m {
        for j := range n {
            if grid[i][j] == '0' {
                continue
            }

            cleanup(i, j)
            islands++
        }
    }

    return islands
}